dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,303 @@ cdef int _local_tracker(
+                copypoint(&directions[i-1, 0], dir)
+
+                if tissue_class[0] != TRACKPOINT:
+                    # The tracking stops. PFT returned a valid stopping point
+                    # (ENDPOINT, OUTSIDEIMAGE) or failed to find one
+                    # (INVALIDPOINT, PYERROR)
+                    break
+            else:
+                # PFT was run more times than `pft_max_trials` without finding
+                # a valid stopping point. The tracking stops with INVALIDPOINT.
+                break
+        else:
+            # The tracking stops with a valid point (ENDPOINT, OUTSIDEIMAGE)
+            # or an invalid point (PYERROR)
+            break
+
+    if tissue_class[0] == OUTSIDEIMAGE or tissue_class[0] == PYERROR:
+        i -= 1
+    return i
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef _pft(np.float_t[:, :] streamline,
+          int streamline_i,
+          np.float_t[:, :] directions,
+          DirectionGetter dg,
+          ConstrainedTissueClassifier tc,
+          double* voxel_size,
+          double step_size,
+          TissueClass * tissue_class,
+          int pft_nbr_steps,
+          int particle_count,
+          np.float_t[:, :, :, :] particle_paths,
+          np.float_t[:, :, :, :] particle_dirs,
+          np.float_t[:] particle_weights,
+          np.int_t[:, :, :] particle_states):
+    cdef:
+        double sum_weights, sum_squared, N_effective, rdm_sample
+        double point[3]
+        double dir[3]
+        double voxdir[3]
+        double eps = np.finfo(np.float).eps
+        int s, p, j
+
+    if pft_nbr_steps <= 0:
+        return streamline_i
+
+    for p in range(particle_count):
+        copypoint(&streamline[streamline_i, 0], &particle_paths[0, p, 0, 0])
+        copypoint(&directions[streamline_i, 0], &particle_dirs[0, p, 0, 0])
+        particle_weights[p] = 1. / particle_count
+        particle_states[0, p, 0] = TRACKPOINT
+        particle_states[0, p, 1] = 0
+
+    for s in range(pft_nbr_steps):
+        for p in range(particle_count):
+            if particle_states[0, p, 0] != TRACKPOINT:
+                for j in range(3):
+                    particle_paths[0, p, s, j] = 0
+                    particle_dirs[0, p, s, j] = 0
+                continue  # move to the next particle
+            copypoint(&particle_paths[0, p, s, 0], point)
+            copypoint(&particle_dirs[0, p, s, 0], dir)
+
+            if dg.get_direction_c(point, dir):
+                particle_states[0, p, 0] = INVALIDPOINT
+                particle_weights[p] = 0
+            else:
+                for j in range(3):
+                    voxdir[j] = dir[j] / voxel_size[j]
+                fixed_step(point, voxdir, step_size)
+                copypoint(point, &particle_paths[0, p, s + 1, 0])
+                copypoint(dir, &particle_dirs[0, p, s + 1, 0])
+                particle_states[0, p, 0] = tc.check_point_c(point)
```

Comment:
Any reason why you shared the `particle tissue_class` with the `particle length`? IMO, it's harder to understand for no real gain. Shouldn't they be 2 arrays?

Reply:
I thought that would be more clear to group them to reduce the number of arrays... Let me try with 2 arrays.